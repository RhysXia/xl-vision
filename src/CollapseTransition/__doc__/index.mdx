# CollapseTransition

可以方便地对于不定高度或者宽度的内容进行折叠或者展开的动画组件。

:::demo 基础用法

通过`CollaseTrannsition`包裹对应的元素来触发折叠和展开的动画效果。通过`transitionClassName`自定义动画方式。如果不指定，将使用默认动画`width/height 1s ease`

```jsx
import React from 'react'
import { CollapseTransition, Button } from 'xl-vision'
export default () => {
  const [active, setActive] = React.useState(false)

  return (
    <div className='wrapper'>
      <Button
        onClick={() => {
          setActive(!active)
        }}
      >
        Click
      </Button>
      <CollapseTransition in={active} transitionClassName={'demo-fade'}>
        <div className='demo'>DEMO</div>
      </CollapseTransition>
    </div>
  )
}
```

```scss
.demo {
  height: 10rem;
  color: #fff;
  font-size: 3rem;
  line-height: 10rem;
  text-align: center;
  vertical-align: middle;
  background-color: #007bff;
}

.wrapper::deep .demo-fade {
  transition: height 2s ease;
}
```

:::

::: demo 水平模式

通过属性`horizontal`在水平方向进行展开和折叠动画

```jsx
import React from 'react'
import { CollapseTransition, Button } from 'xl-vision'
export default () => {
  const [active, setActive] = React.useState(false)

  return (
    <div className='wrapper'>
      <Button
        onClick={() => {
          setActive(!active)
        }}
      >
        Click
      </Button>
      <CollapseTransition in={active} transitionClassName={'demo-fade'} horizontal>
        <div className='demo'>DEMO</div>
      </CollapseTransition>
    </div>
  )
}
```

```scss
.demo {
  height: 10rem;
  width: 100%;
  color: #fff;
  font-size: 3rem;
  line-height: 10rem;
  text-align: center;
  vertical-align: middle;
  background-color: #007bff;
}

.wrapper::deep .demo-fade {
  transition: width 2s ease;
}
```

:::

## CollapseTransition API

| 参数 | 说明 | 类型 | 可选值 | 默认值 | 是否必填 |
| --- | --- | --- | --- | --- | --- |
| in | 触发进出场动画,`true`表示进场，`false`表示出场 | boolean | - | — | true |
| transitionClassName | 指定动画过程中的 class 名,如果不指定，会使用默认动画`height(width) 1s ease` | string | - | - | false |
| children | 子元素 | React.ReactElement | - | — | true |
| horizontal | 是否横向 | boolean | - | — | - |
| transitionOnFirst | 是否初始化完成时就触发动画 | boolean | - | false | false |
| mountOnEnter | 当第一次进场动画时才挂载子节点，见<Link to="Transition">**Transition**</Link> | boolean | - | false | false |
| unmountOnLeave | 当出场动画后就卸载子节点，见<Link to="Transition">**Transition**</Link> | boolean | - | false | false |

