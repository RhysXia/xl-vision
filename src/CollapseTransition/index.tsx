import PropTypes from 'prop-types'
import React from 'react'
import CSSTransition from '../CSSTransition'

export interface CollapseTransitionProp {
  children: React.ReactElement<React.HTMLAttributes<HTMLElement>>
  forceRender?: boolean
  in: boolean
  transitionClassName?: string
  horizontal?: boolean
  transitionOnFirst?: boolean
}

const CollapseTransition: React.FunctionComponent<CollapseTransitionProp> = (props) => {
  const {
    children,
    transitionClassName,
    in: inProp,
    forceRender,
    horizontal,
    transitionOnFirst
  } = props

  const wrapperRef = React.useRef<HTMLDivElement>(null)

  const styles: React.CSSProperties = React.useMemo(() => {
    return {
      boxSizing: 'border-box',
      overflow: 'hidden',
      paddingBottom: '0',
      paddingTop: '0',
      transition: transitionClassName ? '' : `${horizontal ? 'width' : 'height'} 1s ease`
    }
  }, [transitionClassName, horizontal])

  const transitionEvents = React.useMemo(() => {
    const key = horizontal ? 'width' : 'height'
    const offsetKey = horizontal ? 'offsetWidth' : 'offsetHeight'
    let size: number
    let isCancelled = false
    return {
      beforeEnter(el: HTMLElement) {
        if (isCancelled) {
          el.style[key] = el[offsetKey] + 'px'
          isCancelled = false
        } else {
          size = wrapperRef.current![offsetKey]
          el.style[key] = '0'
        }
      },
      enter(el: HTMLElement, _done: () => void, isCancelled: () => boolean) {
        if (!isCancelled()) {
          // 高度设置为内容高度
          el.style[key] = size + 'px'
        }
      },
      afterEnter(el: HTMLElement) {
        // 防止内容高度变更后高度不会自动改变
        el.style[key] = ''
      },
      enterCancelled() {
        isCancelled = true
      },
      beforeLeave(el: HTMLElement) {
        if (isCancelled) {
          isCancelled = false
        } else {
          size = wrapperRef.current![offsetKey]
        }
        el.style[key] = wrapperRef.current![offsetKey] + 'px'
      },
      leave(el: HTMLElement, _done: () => void, isCancelled: () => boolean) {
        if (!isCancelled()) {
          el.style[key] = '0'
        }
      },
      leaveCancelled() {
        isCancelled = true
      }
    }
  }, [horizontal])

  return (
    <CSSTransition
      transitionOnFirst={transitionOnFirst}
      {...transitionEvents}
      in={inProp}
      forceRender={forceRender}
    >
      <div className={transitionClassName} style={styles}>
        <div ref={wrapperRef} style={{ position: 'relative' }}>
          {children}
        </div>
      </div>
    </CSSTransition>
  )
}

CollapseTransition.propTypes = {
  transitionClassName: PropTypes.string,
  horizontal: PropTypes.bool,
  children: PropTypes.element.isRequired,
  in: PropTypes.bool.isRequired,
  forceRender: PropTypes.bool,
  transitionOnFirst: PropTypes.bool
}

export default CollapseTransition
