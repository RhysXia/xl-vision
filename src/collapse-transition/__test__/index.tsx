import { mount } from 'enzyme'
import React from 'react'
import CollapseTransition from '..'

describe('CollapseTransition', () => {
  it('基本使用', async () => {
    const wrapper = mount(
      <CollapseTransition show={false}>
        <div />
      </CollapseTransition>
    )

    expect(wrapper.render()).toMatchSnapshot()

    wrapper.setProps({
      show: true
    })

    wrapper.update()

    expect(wrapper.render()).toMatchSnapshot()
  })
})
