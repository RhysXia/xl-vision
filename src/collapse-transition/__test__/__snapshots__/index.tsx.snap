// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapseTransition 基本使用 1`] = `null`;

exports[`CollapseTransition 基本使用 2`] = `
<div
  style="box-sizing: border-box; overflow: hidden; padding-bottom: 0px; padding-top: 0px; transition: height 1s ease; height: 0px;"
>
  <div
    style="position: relative;"
  >
    <div />
  </div>
</div>
`;
